src/LinuxTracing/LibunwindstackUnwinder.cpp

```diff
@@ -116,7 +125,99 @@ LibunwindstackResult LibunwindstackUnwinderImpl::Unwind(
 
   return LibunwindstackResult{unwinder.ConsumeFrames(), unwinder.LastErrorCode()};
 }
+// This functions detects if a frame pointer register was set in the given program counter using
+// the given Dwarf section.
+// It does so by verifying if "Canonical Frame Address" gets computed immediately from $rbp (with
+// offset 16 to skip the old frame pointer and the return address).
+// The function returns nullopt if the required Dwarf information is not available.
+static std::optional<bool> HasFramePointerSetFromDwarfSection(
+    uint64_t rel_pc, unwindstack::DwarfSection* dwarf_section,
+    std::map<uint64_t, unwindstack::DwarfLocations>* loc_regs_cache) {
+  if (dwarf_section == nullptr) {
+    return false;
+  }
+
+  auto cache_it = loc_regs_cache->upper_bound(rel_pc);
+  if (cache_it == loc_regs_cache->end() || rel_pc < cache_it->second.pc_start) {
+    const unwindstack::DwarfFde* fde = dwarf_section->GetFdeFromPc(rel_pc);
+    if (fde == nullptr) {
+      return std::nullopt;
+    }
+    unwindstack::DwarfLocations loc_regs;
+    if (!dwarf_section->GetCfaLocationInfo(rel_pc, fde, &loc_regs, unwindstack::ARCH_X86_64)) {
+      return std::nullopt;
+    }
+    cache_it = loc_regs_cache->emplace(loc_regs.pc_end, std::move(loc_regs)).first;
+  }
+
+  unwindstack::DwarfLocations* loc_regs = &cache_it->second;
+  auto cfa_entry = loc_regs->find(unwindstack::CFA_REG);
+  if (cfa_entry == loc_regs->end()) {
+    return false;
+  }
+  const unwindstack::DwarfLocation* loc = &cfa_entry->second;
+  ORBIT_CHECK(loc != nullptr);
+  if (loc->type == unwindstack::DWARF_LOCATION_REGISTER) {
+    // From the Dwarf standard:
+    //  "Typically, the CFA is defined to be the value of the stack pointer at the call site in the
+    //  previous frame (which may be different from its value on entry to the current frame)"
+    // So for the frame pointer case, the "value of the stack pointer at the call site" is:
+    // $rbp + 8 (for the prev. frame pointer) + 8 (for the return address)
+    if (loc->values[0] == unwindstack::X86_64_REG_RBP && loc->values[1] == 16) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+std::optional<bool> LibunwindstackUnwinderImpl::HasFramePointerSet(uint64_t instruction_pointer,
+                                                                   unwindstack::Maps* maps) {
+  unwindstack::MapInfo* map_info = maps->Find(instruction_pointer);
+  if (map_info == nullptr) {
+    return std::nullopt;
+  }
+
+  // We don't have enough process memory collected to read the object file from memory (if memory
```

Comment:
I _think_ that the process memory implementation would support reading from process memory even for parts that are not in the collected stack, however it is slow if it's outside of the collected bounds. In my experiments in the unwinder, it was slower than reading from file.

Reply:
> Well, from the setting, we will never have enough memory to do so. And I don't think it is worth passing the additional arguments to here, just to internally fall back to file reading.

It wouldn't be the memory from the stack dump. You would use `Memory::CreateProcessMemory` or `Memory::CreateProcessMemoryCached`. We also use that in `StackAndProcessMemory`. Maybe not important for us, but with that libunwindstack also supports ELFs that are only in memory. This is again what happens with `StackAndProcessMemory`, which for example allows us to support vDSO.
I also wanted to suggest that it might also make it faster than reading from file, but if it's the opposite (`process_vm_readv` used internally is slow?), then maybe we don't want it for that reason.
Either way, the reason in the comment doesn't really hold.